lib/puppet/resource_api/type_definition.rb

```diff
@@ -35,4 +35,56 @@ def feature?(feature)
     end
     supported
   end
+
+  # validates a resource hash against its type schema
+  def check_schema(resource)
+    namevars.each do |namevar|
+      if resource[namevar].nil?
+        raise Puppet::ResourceError, "`#{name}.get` did not return a value for the `#{namevar}` namevar attribute"
+      end
+    end
+
+    message = "Provider returned data that does not match the Type Schema for `#{name}[#{resource[namevars.first]}]`"
+
+    rejected_keys = check_schema_keys(resource) # removes bad keys
+    bad_values = check_schema_values(resource.dup)
```

Comment:
Is `check_schema` allowed to modify `resource` or not? If yes, do you need to `dup` here? If not, shouldn't this `dup` before `check_schema_keys`?

Also, now that the question has come up, you _might_ want to add a unit test making sure that `resource` doesn't get modified when checking.